src/tracker/tracker_sampler.jl

```diff
@@ -0,0 +1,100 @@
+abstract type AbstractROI end
+mutable struct BoxROI{T <: AbstractArray, S <: AbstractArray} <: AbstractROI
+    img::T
+    bound::S
+end
+
+abstract type AbstractTrackerSampler end
+mutable struct CurrentSampler{F <: AbstractFloat} <: AbstractTrackerSampler
+    overlap::F
+    search_factor::F
+    mode::S where S <: Symbol
+    sampling_region::SVector{4, Integer}
+    valid_region::SVector{4, Integer}
+    tracked_patch::SVector{4, Integer}
+    CurrentSampler{F}(overlap, search_factor) where{F <: AbstractFloat} = new(overlap, search_factor)
+end
+
+CurrentSampler(overlap::F = 0.99, search_factor::F = 1.8) where{F <: AbstractFloat} = CurrentSampler{F}(overlap, search_factor)
+
+function sample_roi(sampler::CurrentSampler, box::BoxROI)
+    sampler.tracked_patch = box.bound
+    sampler.valid_region = SVector{4, Integer}([1, 1, size(box.img, 1), size(box.img, 2)])
+
+    height = box.bound[3] - box.bound[1] + 1
+    width = box.bound[4] - box.bound[2] + 1
+
+    sampling_region_min_y = max(0, floor(Integer, box.bound[1] - height*((sampler.search_factor - 1) / 2) + 1))
+    sampling_region_min_x = max(0, floor(Integer, box.bound[2] - width*((sampler.search_factor - 1) / 2) + 1))
+
+    sampling_region_height = min(floor(Integer, height*sampler.search_factor), size(box.img, 1) - sampling_region_min_y + 1)
+    sampling_region_width = min(floor(Integer, width*sampler.search_factor), size(box.img, 2) - sampling_region_min_x + 1)
+
+    sampling_region_max_y = sampling_region_min_y + sampling_region_height - 1
+    sampling_region_max_x = sampling_region_min_x + sampling_region_width - 1
+
+    sampling_region = SVector{4, Integer}([sampling_region_min_y, sampling_region_min_x, sampling_region_max_y, sampling_region_max_x])
```

Comment:
This is not a good idea. You are allocating an array by invoking the square brackets `[..]`, only to immediately create another array on the stack. You don't need the square brackets here. Perhaps this ought to just be a range objects anyway? i.e.  `sampling_region_min_y:sampling_region_max_y` and `sampling_region_min_x:sampling_region_max_x`.